application/ledger_query_service/src/ethereum/queries.rs

```diff
@@ -27,11 +27,42 @@ pub struct TransactionQuery {
     transaction_data_length: Option<usize>,
 }
 
-type Topics = Vec<H256>;
+type Topic = H256;
 
 #[derive(Debug, Clone, Serialize, Deserialize, Default)]
-pub struct LogQuery {
-    topics: Vec<Topics>,
+pub struct EventQuery {
+    event_matchers: Vec<EventMatcher>,
+}
+
+// Event Matcher work similar as web3 filters:
+// https://web3js.readthedocs.io/en/1.0/web3-eth-subscribe.html?highlight=filter#subscribe-logs
+// E.g. this EventMatcher would match this Log:
+// EventMatcher {
+// address: 0xe46FB33e4DB653De84cB0E0E8b810A6c4cD39d59,
+// data: None,
+// topics: [
+// None,
+// Some(0x000000000000000000000000e46fb33e4db653de84cb0e0e8b810a6c4cd39d59),
+// None()
+// ],
+//
+// Log:
+// [ { address: '0xe46FB33e4DB653De84cB0E0E8b810A6c4cD39d59',
+// data: '0x123',
+// ..
+// topics:
+// [ '0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef',
+// '0x000000000000000000000000e46fb33e4db653de84cb0e0e8b810a6c4cd39d59',
+// '0x000000000000000000000000d51ecee7414c4445534f74208538683702cbb3e4' ],
+// },
+// .. ] //Other data omitted
+// }
+
+#[derive(Debug, Clone, Serialize, Deserialize, Default)]
+struct EventMatcher {
+    address: Option<Address>,
+    data: Option<Bytes>,
+    topics: Vec<Option<Topic>>,
```

Comment:
read the comment above ;)
you need to have an option because you might want to ignore some topics in the list. If you want to know all transactions which called `transfer` with the `toAddress = something` but you don't care where it comes from, i.e. `fromAddress = ignored` than you would need this feature.

Reply:
I think I get it. Is it because the topics are indexed? In that case I would have used a `Vec<(u32, Topic)>` but maybe this is fine or even better :stuck_out_tongue:

Reply:
true, that would have been another option.